### PR TITLE
fix: webpack-dev-server re-run only updated

### LINKED ISSUE #16253: [Component tests] As soon as I change a test file I get "Your tests are loading..."
### Current behavior

I have some component tests. The component test runner runs them fine. When I change a component test file, I expect the test runner to detect the change and run the tests again. It doesn't do that. When I refresh the component test runner manually, I get "Your tests are loading..." forever.

**At this point I have to kill the component test runner and start it again.**

### Desired behavior

I expect either one of these behavior, from most favorite to least favorite:

1. The best would be that the component test runner detects changes to my component test file and refreshes itself;
2. If that is not possible, I expect to be able to refresh the component test runner manually and get the changed test executed.

### Versions

- Cypress 7.2.0. It was working fine in Cypress 6.8.0.
- Browser: Chromium 89.0.4389.90
- OS: Fedora 33

### LINKED ISSUE #17151: running component tests without component directory fails with infinite recursion
### Current behavior
When running the cypress component tests without a cypress/component directory it appears to try and recurse in cypress/component directories. The dev server then finally gives up with:

```
errno: -36,
  code: 'ENAMETOOLONG',
  syscall: 'stat',
  path: 'cypress/component/cypress/component/cypress/component/cypress/component/cypress/component/cypress/component/cypress/component/cypress/component/cypress/component/.../cypress/component'
```

This is easily resolved by adding a component directory.

### Desired behavior
Some indication/warning that the component directory doesn't exist.

### Test code to reproduce
This happens with the simplest example I could find.

### Versions
```
    "@babel/core": "^7.10.4",
    "webpack-dev-server": "^3.11.2",
    "@svgr/webpack": "^5.5.0",
    "@cypress/webpack-dev-server": "^1.3.1",
    "html-webpack-plugin": "4.0.0",
    "css-loader": "^5.2.6",
    "file-loader": "^4.2.0"
```

 ../../node_modules/.bin/cypress --version
Cypress package version: 7.5.0
Cypress binary version: 7.5.0
Electron version: 12.0.0-beta.14
Bundled Node version: 14.15.1

### PR DESCRIPTION
**Pull request by elevatebart**
- Closes #17151
- Closes #16253

### User facing changelog
When a spec file or one of its dependencies is updated, it should re-run if it is active.
Before this PR any active spec would be re-run when saving any file.

### CODE REVIEW

File: npm/webpack-dev-server/test/e2e.spec.ts [hunk 1]
```diff
@@ -87,9 +101,13 @@ describe('#startDevServer', () => {
       },
     })
 
-    return new Promise((res) => {
-      devServerEvents.on('dev-server:compile:success', () => {
-        close(() => res())
+    return new Promise<void>((res) => {
```
Comment (lmiller1990) [inline npm/webpack-dev-server/test/e2e.spec.ts:104]: What does this change do? I don't understand the purpose/what this is testing.

File: npm/vue/package.json [hunk 2]
```diff
@@ -30,7 +30,7 @@
     "@rollup/plugin-node-resolve": "^11.1.1",
     "@vue/cli-plugin-babel": "~4.4.0",
     "@vue/cli-service": "~4.4.0",
-    "@vue/compiler-sfc": "3.1.5",
+    "@vue/compiler-sfc": "3.2.0",
```
Comment (lmiller1990) [inline npm/vue/package.json:33]: I am not sure we should update deps in master prior to develop - unless the rest of the packages are already on Vue 3.2?

File: npm/webpack-dev-server/src/plugin.ts [hunk 3]
```diff
@@ -85,12 +86,10 @@ export default class CypressCTOptionsPlugin {
...
 
-    compiler.hooks.afterEmit.tap(
+    compiler.hooks.done.tap(
       'CypressCTOptionsPlugin',
-      (compilation) => {
-        if (!compilation.getStats().hasErrors()) {
-          this.devServerEvents.emit('dev-server:compile:success')
-        }
+      () => {
+        this.refreshCompile = true
```
Comment (lmiller1990) [inline npm/webpack-dev-server/src/plugin.ts:92]: Is it correct to trigger this even if there is errors? I notice we deleted `compilation.getStats().hasErrors()`

**Review by lmiller1990 — COMMENTED** [review]
Some questions/comments. I'll give this one a test soon.

File: npm/vue/package.json (hunk 2 above)
Comment (elevatebart) [reply npm/vue/package.json:33]: agreed, I'll roll them back

File: npm/webpack-dev-server/src/plugin.ts (hunk 3 above)
Comment (elevatebart) [reply npm/webpack-dev-server/src/plugin.ts:92]: I truly have no idea. 
The UX is weird when compilation fails as the tests look green but an error shows in the AUT panel.

### webpack in master
the command log pannel seems untouched ebven if the build failed
Webpack error on master

### webpack in this PR
command log is sending a cryptic set of errors
Webpack error on this PR

Comment (elevatebart) [reply npm/webpack-dev-server/src/plugin.ts:92]: reversed this change in c8f5816

File: npm/webpack-dev-server/test/e2e.spec.ts (hunk 1 above)
Comment (ImCesar) [reply npm/webpack-dev-server/test/e2e.spec.ts:104]: If I'm understanding correctly before this change the `dev-server:compile:success` event would fire after compiling and outputting the compiled files. So this test would make sure after the dev server was done starting up we got that event. 

With the proposed change it is now firing after a successful rebuild and if the changed file is one of the specs we're watching. So before we can run the old test we want to modify a file to trigger a rebuild. 

If that's a correct description of what we're doing, do we want a test where we test touching a file that isn't being watched?

**Comment by lmiller1990** [comment]
@elevatebart This is not have the expected effect for me. I ran `yarn install`, `cd packages/runner-ct` and opened `SpecList.spec.tsx`. I then edited `ScriptError.spec.tsx`, but the runner still re-ran `SpecList.spec.tsx`.

**Comment by elevatebart** [comment]
Hey @lmiller1990 Thank you for the sanity check I totally had skipped webpack 4 compat. 
I updated the process and tested in both wp 4 and wp 5. it should be good to go.

**Comment by lmiller1990** [comment]
Something is not right here. I was able to break the runner easily. Here's a video. The `RunnerCt` component renders a `SpecList`. When I changed `SpecList`, the runner basically freezes.

This does not happen in develop.

https://user-images.githubusercontent.com/19196536/134097132-ebf5b65b-bb57-4f8f-850a-a52e10e3e3cd.mov

**Review by lmiller1990 — CHANGES_REQUESTED** [review]
Left comment with a bug I found.

**Comment by lmiller1990** [comment]
@elevatebart should we close this one off and revisit in the future? Seems progress has stalled (just going through and cleaning up some old PRs).

**Pull request closed without merging**